feat: maintenance mode

insertish opened the pull request:
Issue: https://github.com/immich-app/yucca/issues/8
Topic: https://discord.com/channels/979116623879368755/1433785614205194361

This PR adds:
- Maintenance mode which can be enabled from settings.
- CLI to toggle maintenance mode.
- Separate "maintenance worker" which replaces API/microservices and puts Immich in a state where it can perform database restores, etc.

## How Has This Been Tested?

<!-- Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration -->

- [x] Service unit tests for CLI (enable, disable maintenance mode)
- [x] Service unit tests for maintenance service (query status, login, start, end, create login URL)
- [x] Service unit tests for maintenance worker service (system config, log secret, authenticate, login, start, end)
- [x] e2e integration test for server (consume maintenance API, enable maintenance, exit maintenance)
- [x] e2e integration test for web (enable and disable maintenance through settings, redirect users to/back) 

<h2>Demo</h2>

[Screencast_20251110_095421.webm](https://github.com/user-attachments/assets/07083951-b2c5-4a7e-a77d-d71a7bc6a1ee)

<details><summary><h2>Screenshots</h2></summary>

| User interface when logged in and in maintenance mode | User interface when logged out and in maintenance mode |
|---|---|
| <img width="1491" height="1157" alt="image" src="https://github.com/user-attachments/assets/85aa1fe4-11af-4326-84e4-5c6b136d0b42" /> | <img width="1221" height="848" alt="image" src="https://github.com/user-attachments/assets/5117f937-ebca-4512-b993-f811e17ea63e" /> |

Mobile app gracefully fails and displays an error in the corner:

| <img width="1080" height="1900" alt="image" src="https://github.com/user-attachments/assets/90126283-62ec-4823-92f5-3966fe2722ee" /> | <img width="1080" height="1900" alt="image" src="https://github.com/user-attachments/assets/9f36b7d7-585a-46d9-b3dd-1602593bc76b" /> | <img width="1080" height="1900" alt="image" src="https://github.com/user-attachments/assets/1427fac4-efc7-472f-a0ac-cefb91ec99a9" /> |
|---|---|---|

</details>

## API Changes

- `GET /server/config` route now includes `maintenanceMode: boolean`
- `POST /admin/maintenance/login`: login with token generated when maintenance mode is enabled
- `POST /admin/maintenance/start`: switch Immich into maintenance mode (requires Maintenance permission)
- `POST /admin/maintenance/end`: bring Immich out of maintenance mode (requires maintenance auth)

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation if applicable
- [x] I have no unrelated changes in the PR.
- [x] I have confirmed that any new dependencies are strictly necessary. \
    (*adds jose for jwt validation/signing*)
- [x] I have written tests for new code (if applicable)
- [x] I have followed naming conventions/patterns in the surrounding code
- [x] All code in `src/services/` uses repositories implementations for database calls, filesystem operations, etc.
- [x] All code in `src/repositories/` is pretty basic/simple and does not have any immich specific logic (that belongs in `src/services/`)

## Please describe to which degree, if any, an LLM was used in creating this pull request.

No code generation.

## Further Work (out of scope)

- With the introduction of `enums/ExitCode`, all existing `process.exit` references should probably be adapted to use this.

insertish committed:
feat: add a `maintenance.enabled` config flag

insertish committed:
feat: implement graceful restart
feat: restart when maintenance config is toggled

insertish committed:
feat: boot a stripped down maintenance api if enabled

insertish committed:
feat: cli command to toggle maintenance mode

insertish committed:
chore: fallback IMMICH_SERVER_URL environment variable in process

insertish committed:
chore: add additional routes to maintenance controller

insertish committed:
fix: don't wait for nest application to close to finish request response

insertish committed:
chore: add a failsafe on restart to prevent other exit codes from preventing restart

insertish committed:
feat: redirect into/from maintenance page

insertish committed:
refactor: use system metadata for maintenance status

insertish committed:
refactor: wait on WebSocket connection to refresh

insertish committed:
feat: broadcast websocket event on server restart
refactor: listen to WS instead of polling

insertish committed:
refactor: bubble up maintenance information instead of hijacking in fetch function
feat: show modal when server is restarting

insertish committed:
chore: increase timeout for ungraceful restart

insertish committed:
refactor: deduplicate code between api/maintenance workers

insertish committed:
fix: skip config check if database is not initialised

insertish committed:
fix: add `maintenanceMode` field to system config test

insertish committed:
refactor: move maintenance resolution code to static method in service

insertish committed:
chore: clean up linter issues

insertish committed:
chore: generate dart openapi

insertish committed:
refactor: use try{} block for maintenance mode check

insertish committed:
fix: logic error in server redirect

insertish committed:
chore: include `maintenanceMode` key in e2e test

insertish committed:
chore: add i18n entries for maintenance screens

insertish committed:
chore: remove negated condition from hook

insertish committed:
fix: should set default value not override in service

insertish committed:
fix: minor error in page

insertish committed:
feat: initial draft of maintenance module, repo., worker controller, worker service

insertish committed:
refactor: move broadcast code into notification service

insertish committed:
chore: connect websocket on client if in maintenance

insertish committed:
chore: set maintenance module app name

insertish committed:
refactor: rename repository to include worker
chore: configure websocket adapter

insertish committed:
feat: reimplement maintenance mode exit with new module

insertish committed:
refactor: add a constant enum for ExitCode

insertish committed:
refactor: remove redundant route for maintenance

insertish committed:
refactor: only spin up kysely on boot (rather than a Nest app)

insertish committed:
refactor(web): move redirect logic into +layout file where modal is setup

insertish committed:
feat: add Maintenance permission

insertish committed:
refactor: merge common code between api/maintenance

insertish committed:
fix: propagate changes from the CLI to servers

insertish committed:
feat: maintenance authentication guard

insertish committed:
refactor: unify maintenance code into repository
feat: add a step to generate maintenance mode token

insertish committed:
feat: jwt auth for maintenance

insertish committed:
refactor: switch from nest jwt to just jsonwebtokens

insertish committed:
feat: log into maintenance mode from CLI command

insertish committed:
refactor: use `secret` instead of `token` in jwt terminology
chore: log maintenance mode login URL on boot
chore: don't make CLI actions reload if already in target state

insertish committed:
docs: initial draft for maintenance mode page

insertish committed:
refactor: always validate the maintenance auth on the server

insertish committed:
feat: add a link to maintenance mode documentation

insertish committed:
feat: redirect users back to the last page they were on when exiting maintenance

insertish committed:
refactor: provide closeFn in both maintenance repos.

insertish committed:
refactor: ensure the user is also redirected by the server

insertish committed:
chore: swap jsonwebtoken for jose

insertish committed:
refactor: introduce AppRestartEvent w/o secret passing

insertish committed:
refactor: use navigation goto

insertish committed:
refactor: use `continue` instead of `next`

insertish committed:
chore: lint fixes for server

insertish committed:
chore: lint fixes for web

insertish committed:
test: add mock for maintenance repository

insertish committed:
test: add base service dependency to maintenance

insertish committed:
chore: remove @types/jsonwebtoken

jrasm91 commented on server/src/middleware/maintenance-auth.guard.ts:
```diff
@@ -0,0 +1,58 @@
+};
+
+export interface MaintenanceAuthRequest extends Request {
+  maintenanceAuth?: MaintenanceAuthDto;
```
can we just call this `auth`?

jrasm91 commented on server/src/workers/maintenance.ts:
```diff
@@ -0,0 +1,31 @@
+    ssr: false,
+  });
+
+  void maintenanceWorkerRepository.logSecret();
```
I would probably prefer to call a service method here.

jrasm91 commented on web/src/routes/maintenance/+page.svelte:
```diff
@@ -0,0 +1,44 @@
+              <a href="https://docs.immich.app/administration/maintenance-mode" target="_blank" rel="noopener">
+                {message}
+              </a>
+            </span>
```
Just use the `Link` component from `@immich/ui`. It still automatically add target and rel if it is an external url.

jrasm91 commented on web/src/routes/+layout.svelte:
```diff
@@ -104,6 +107,32 @@
+      return;
+    }
+
+    if (isRestarting.isMaintenanceMode !== page.url.pathname.startsWith(AppRoute.MAINTENANCE)) {
```
You have this maintenance route check three times now, let's put it into a utility method like `isMaintenanceRoute` or something.

jrasm91 commented on web/src/routes/+layout.svelte:
```diff
@@ -104,6 +107,32 @@
+              ? maintenanceCreateUrl(new URL(location.href))
+              : maintenanceReturnUrl(new URLSearchParams(location.search), location.origin),
+            { invalidateAll: true },
+          );
```
Maybe we should just try to refresh the page, which should handle redirecting to the maintenance page

insertish committed:
refactor: close database connection after startup check

insertish committed:
refactor: use `request#auth` key

insertish committed:
refactor: use service instead of repository
chore: read token from cookie if possible
chore: rename client event to AppRestartV1

insertish committed:
refactor: more concise redirect logic on web

insertish committed:
refactor: move redirect check into utils
refactor: update translation strings to be more sensible

insertish committed:
refactor: always validate login (i.e. check cookie)

insertish committed:
refactor: lint, open-api, remove old dto

insertish committed:
refactor: encode at point of usage

insertish committed:
refactor: remove business logic from repositories

insertish committed:
chore: fix server/web lints

insertish commented on server/src/repositories/maintenance-worker.repository.ts:
```diff
@@ -0,0 +1,152 @@
+  constructor(
+    private logger: LoggingRepository,
+    private configRepository: ConfigRepository,
+    private systemMetadataRepository: SystemMetadataRepository,
```
It appears you can't get the Nest application directly through ModuleRef, I can't find any workaround to this.

insertish committed:
refactor: remove repository mock

insertish committed:
chore: fix formatting

insertish committed:
test: write service mocks for maintenance mode

insertish committed:
test: write cli service tests

insertish committed:
fix: catch errors when closing app

insertish committed:
fix: always report no maintenance when usual API is available

insertish committed:
test: api e2e maintenance spec

insertish committed:
chore: add response builder

insertish committed:
chore: add helper to set maint. auth cookie

insertish committed:
feat: add SSR to maintenance API

insertish committed:
test(e2e): write web spec for maintenance

insertish committed:
chore: clean up lint issues

insertish committed:
chore: format files

insertish committed:
feat: perform 302 redirect at server level during maintenance

insertish committed:
fix: keep trying to stop immich until it succeeds (CLI issue)

insertish committed:
chore: lint/format

insertish committed:
refactor: annotate references to other services in worker service

insertish committed:
chore: lint

danieldietzler commented on server/src/services/cli.service.ts:
```diff
@@ -38,6 +42,63 @@ export class CliService extends BaseService {
+
+    if (state.isMaintenanceMode) {
+      return {
+        authUrl: await MaintenanceService.createLoginUrl(baseUrl, payload, state.secret),
```
Services shouldn't call out to other services. If this serves more of like a helper function, move it to some util file. If it's trivial, just inline it.

insertish replied:
Moved to a utility function

danieldietzler commented on server/src/services/maintenance.service.ts:
```diff
@@ -0,0 +1,84 @@
+      }
+
+      return state.secret;
+    });
```
I would personally move this block inside the .then out of there (and not try to cram it into a single-line assignment. So more of like

```ts
const { isMaintenanceMode, secret: currentSecret } = await this.getMaintenanceMode();

if (!isMaintenanceMode) {
	throw new Error(...)
} 

return await MaintenanceService.createLoginUrl(baseUrl, auth, secret ?? currentSecret);
```

insertish replied:
Refactored out but instead reassigning `secret` as I'm only running the state fetch if `secret` is not set.

danieldietzler commented on web/src/lib/utils/server.ts:
```diff
@@ -1,17 +1,17 @@
   defaults.fetch = fetch;
   await initLanguage();
-  await retrieveServerConfig();
+  return await retrieveServerConfig();
```
```suggestion
  return retrieveServerConfig();
```

danieldietzler commented on web/vite.config.ts:
```diff
@@ -6,8 +6,10 @@ import path from 'node:path';
+
 const upstream = {
-  target: process.env.IMMICH_SERVER_URL || 'http://immich-server:2283/',
+  target: process.env.IMMICH_SERVER_URL,
```
Why that change?

insertish commented on server/src/services/maintenance-worker.service.ts:
```diff
@@ -0,0 +1,159 @@
+  /**
+   * {@link _ApiService.ssr}
+   */
+  ssr(excludePaths: string[]) {
```
I quickly realised that the web app would just serve a 404 if this was missing, I'd say it might be deceptively named because the "ssr code" seems to actually serve the app assets.

insertish commented on server/src/controllers/maintenance-worker.controller.ts:
```diff
@@ -0,0 +1,43 @@
+  @MaintenanceRoute()
+  async endMaintenance(): Promise<void> {
+    await this.service.endMaintenance();
+  }
```
Yes but they use different logic/services.
Worker controller exclusively runs in maintenance worker.
Non-worker controller exclusively runs in API worker.

insertish committed:
refactor: move static methods into util

insertish committed:
refactor: assert secret exists in maintenance worker

insertish committed:
refactor: remove assertion which isn't necessary anymore

insertish committed:
refactor: remove assertion

insertish committed:
refactor: remove outer try {} catch block from loadMaintenanceAuth

insertish commented on web/src/lib/utils/maintenance.ts:
```diff
@@ -0,0 +1,41 @@
+}
+
+export const loadMaintenanceAuth = async () => {
+  try {
```
Removed.
I was trying to copy how loadUser/auth.ts was designed but seemingly ignored anything outside of the core block of code.

insertish committed:
refactor: remove unnecessary await

Co-authored-by: Daniel Dietzler <36593685+danieldietzler@users.noreply.github.com>

insertish committed:
refactor: undo earlier change to vite.config.ts

insertish commented on server/src/repositories/maintenance.repository.ts:
```diff
@@ -0,0 +1,64 @@
+import { AppRestartEvent } from 'src/repositories/event.repository';
+
+@Injectable()
+export class MaintenanceRepository {
```
I've removed the code that calls out to another repository, but I'm not sure what to do about `closeFn` or whether it's fine to keep here. (in a previous iteration this did live in a service, but it makes sense to mock it for the service tests)

This repository could also be renamed to something else... `AppShutdownRepository`? `RestartRepository`?

insertish committed:
chore: update tests due to refactors

insertish committed:
revert: vite.config.ts

insertish committed:
test: expect string jwt

insertish committed:
merge: remote-tracking branch 'origin/main' into feat/maintenance-mode

insertish committed:
chore: move blanket exceptions into controllers

insertish committed:
test: update tests according with last change

jrasm91 commented on server/src/main.ts:
```diff
@@ -1,61 +1,145 @@
+  /**
+   * Currently running workers
+   */
+  workers: Partial<Record<ImmichWorker, Worker | ChildProcess>>;
```
```suggestion
  workers: Partial<Record<ImmichWorker, Worker | ChildProcess>> = {};
```

It also might be better to have this be a `Record<ImmichWorker, { kill: () => void }>` to avoid casting it later. Then, just do something like
```typescript
this.workers[name] = () => worker.kill('SIGTERM');
```

jrasm91 commented on server/src/middleware/maintenance-auth.guard.ts:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    const request = context.switchToHttp().getRequest<MaintenanceAuthRequest>();
+    request['auth'] = await this.service.authenticate(request.headers);
```
```suggestion
    request.auth = await this.service.authenticate(request.headers);
```

jrasm91 commented on server/src/repositories/maintenance.repository.ts:
```diff
@@ -0,0 +1,22 @@
+export class MaintenanceRepository {
+  private closeFn?: () => Promise<void>;
+
+  constructor() {}
```
You can just delete this line.

insertish committed:
Merge remote-tracking branch 'origin/main' into feat/maintenance-mode

insertish commented on server/src/repositories/maintenance-worker.repository.ts:
```diff
@@ -0,0 +1,59 @@
+
+  restartApp(state: AppRestartEvent) {
+    // => corresponds to notification.service.ts#onAppRestart
+    this.websocketServer?.emit('AppRestartV1', state, () => {
```
This waits for acknowledgement before continuing.
https://socket.io/docs/v4/emit-cheatsheet/#acknowledgements

insertish committed:
refactor: use respondWithCookie
refactor: merge start/end into one route
refactor: rename MaintenanceRepository to AppRepository
chore: use new ApiTag/Endpoint
refactor: apply other requested changes

insertish committed:
chore: regenerate openapi

insertish committed:
chore: lint/format

insertish committed:
chore: remove secureOnly for maint. cookie

insertish committed:
refactor: move maintenance worker code into src/maintenance\nfix: various test fixes

insertish committed:
refactor: use `action` property for setting maint. mode

insertish committed:
refactor: remove Websocket#restartApp in favour of individual methods

insertish committed:
chore: incomplete commit

insertish committed:
chore: remove stray log

insertish committed:
fix: call exitApp from maintenance worker on exit

insertish committed:
fix: add app repository mock

insertish committed:
fix: ensure maintenance cookies are secure

insertish committed:
fix: run playwright tests over secure context (localhost)

insertish committed:
merge: remote-tracking branch 'origin/main' into feat/maintenance-mode

insertish committed:
test: update other references to 127.0.0.1

insertish committed:
refactor: use serverSideEmitWithAck

insertish committed:
chore: correct the logic in tryTerminate

insertish committed:
test: juggle cookies ourselves

jrasm91 reviewed (approved):
Code looks great!

insertish committed:
chore: fix lint error for e2e spec

insertish committed:
chore: format e2e test

insertish committed:
merge: remote-tracking branch 'origin/main' into feat/maintenance-mode

insertish committed:
fix: set cookie secure/non-secure depending on context

insertish committed:
chore: format files

alextran1502 reviewed (approved):
LGTM, feel free to merge when ready
